1.12.2/1.10.2 Forge: Beyond/Skyfactory3/SevTech Ages crashes: error: StringError("failed to fill whole buffer") / java.util.zip.DataFormatException: incorrect header check

SevTech_Ages_Server_3.1.1 from https://minecraft.curseforge.com/projects/sevtech-ages/files/2686905#additional-files

cargo run --release -- -s localhost:25565 -n

Client crashes on handling FML|HS packet discriminator 2:

thread 'main' panicked at 'Err: IOError(Custom { kind: UnexpectedEof, error: StringError("failed to fill whole buffer") })', src/server/mod.rs:449:33

Forge 1.12.2 with ironchest mod worked in https://github.com/iceiix/stevenarella/pull/144

<details>
<summary>ModList:</summary>

'\x02\xf8\x01\tminecraft\x061.12.2\x03mcp\x049.42\x03FML\t8.0.99.99\x05forge\x0c14.23.5.2815\tivtoolkit\n1.3.3-1.12\rmicdoodlecore\x00\x0esmoothfontcore\rmc1.12.2-1.16\rbnbgamingcore\x060.10.0\x0bfoamfixcore\x057.7.4\nmovillages\x051.5.4\rcharcoalblock\x031.1\x0fadvancedmortars\r1.12.2-1.6.22\x0ccrafttweaker\x064.1.14\x05mtlib\x053.0.5\nmodtweaker\x064.0.16\x03jei\n4.15.0.268\x0cabyssalcraft\x071.9.5.2\x06chisel\x11MC1.12.2-0.2.1.35\x06mantle\r1.12-1.3.3.42\x0etwilightforest\x073.8.689\ntconstruct\x111.12.2-2.12.0.115\racintegration\x051.7.0\tfastbench\x051.6.1\x11actuallyadditions\x0b1.12.2-r146\x07baubles\x051.5.2\x0factuallybaubles\x031.1\tanimalium\x050.3.8\x0cantiqueatlas\x054.5.0\tapplecore\x053.2.0\x04base\x063.12.0\x0econtenttweaker\x0c1.12.2-4.9.1\x14immersiveengineering\x070.12-89\x08geolosys\x052.1.4\x0bcyclicmagic\x071.17.16\x0credstoneflux\x052.1.0\x08mekanism\x101.12.2-9.6.0.351\x06natura\x0f1.12.2-4.3.2.66\x08traverse\x051.6.0\x0ebetterwithmods\x101.12-2.3.20-1027\tappleskin\x051.0.9\x13appliedenergistics2\x0crv6-stable-6\x0earmoreablemobs\x051.1.2\raroma1997core\x072.0.0.2\x0baromabackup\x072.1.1.4\x13aromabackuprecovery\x072.1.1.4\x0castikorcarts\x0e1.12.2-0.1.2.6\rastralsorcery\x071.10.11\x05quark\x08r1.5-146\nautoreglib\x061.3-26\x05bdlib\t1.14.3.12\x0cbedrockbgone\x055.0.8\x10betterwithaddons\x0b0.47-hotfix\x13betterbuilderswands\x060.11.1\rbetterwithlib\n${version}\x0bbibliocraft\x052.4.5\x05waila\x061.8.26\x0emodularrouters\x0c1.12.2-3.2.1\x08guideapi\r1.12-2.1.8-63\nbloodmagic\x101.12.2-2.4.0-102\x0cbnbgaminglib\x062.17.5\x0bbonsaitrees\x051.1.2\tbookshelf\x072.3.574\rbuildcraftlib\x077.99.22\x0ebuildcraftcore\x077.99.22\x12buildcraftbuilders\x077.99.22\x11buildcraftfactory\x077.99.22\x12buildcraftrobotics\x077.99.22\x13buildcrafttransport\x077.99.22\x11buildcraftsilicon\x077.99.22\x0fbuildinggadgets\x052.6.6\x07caliper\x061.1.41\x03car\x061.2.12\ngamestages\x072.0.112\x07carryon\x061.12.1\x0ccd4017be_lib\x056.3.3\x08ceramics\n1.12-1.3.6\tchameleon\n1.12-4.1.3\x08chargers\x071.2.0.4\x0echiselsandbits\x0514.30\x06clumps\x053.1.2\x0ecodechickenlib\t3.2.2.353\x0bcyclopscore\x051.0.5\x12commoncapabilities\x052.0.4\x0erefinedstorage\x061.6.12\x10compactmachines3\x063.0.15\x06conarm\x051.2.3\x14cookingforblockheads\x066.4.69\x0fcrafttweakerjei\x052.0.2\x08cucumber\x051.1.3\tdarkutils\x071.8.223\rdeath_compass\x050.0.3\x12despawningspawners\x031.1\x12dimensionalcontrol\x062.10.3\tdimstages\x062.0.23\x08dungpipe\x031.0\nelevatorid\x051.3.8\temberroot\x051.3.9\x0cenderstorage\t2.4.5.135\x0eenderutilities\x060.7.12\x0bvalkyrielib\x0f1.12.2-2.0.19.1\x11environmentaltech\x0f1.12.2-2.0.19.1\x10extendedcrafting\x051.5.2\x10galacticraftcore\t4.0.1.181\x13galacticraftplanets\t4.0.1.181\rmjrlegendslib\x0c1.12.2-1.1.8\x0cextraplanets\x0c1.12.2-0.5.0\x14farmingforblockheads\x063.1.26\x07farseek\x052.3.1\x0bfastfurnace\x051.2.1\x07fat_cat\x050.0.5\x11ferdinandsflowers\x081.10.4.1\x06findme\x051.1.0\x07foamfix\r0.10.3-1.12.2\x08forgelin\x051.8.2\x11forgemultipartcbe\x082.6.1.81\rmicroblockcbe\x082.6.1.81\x15minecraftmultipartcbe\x082.6.1.81\x13galacticrafttweaker\x0c1.12.2-1.0.3\radvgenerators\t0.9.20.12\ngobblecore\r1.12-0.1.6.35\x07harvest\r1.12-1.2.7-20\nhorsepower\x052.6.2\nhuntingdim\x061.0.36\x06igwmod\x081.4.4-15\x0bmcjtylib_ng\x053.1.1\x08immcraft\x051.5.1\x12immersivepetroleum\x051.1.9\rimmersivetech\x061.3.10\x11improvedbackpacks\x0e1.12.2-1.4.0.0\tincontrol\x053.9.4\x06indlog\x051.3.3\x0cteslacorelib\x061.0.15\x13industrialforegoing\r1.12.2-1.12.2\x0finfoaccessories\x061.0.11\x12integrateddynamics\x051.0.7\x18integrateddynamicscompat\x051.0.0\x0finventorytweaks\x141.64+dev.146.2180b27\rironbackpacks\x0f1.12.2-3.0.8-12\tironchest\x111.12.2-7.0.59.842\x0cironjetpacks\x051.1.0\nitemstages\x062.0.46\njourneymap\x0c1.12.2-5.5.4\x04jarm\x051.1.2\x04jaff\x0c1.7_for_1.12\x04jeid\x081.0.2-26\tkleeslabs\x065.4.11\x03lex\x051.0.8\tlttweaker\x061.1.14\x0emagma_monsters\x050.3.0\nmekatweaks\x071.0.0-3\tmercurius\x051.0.6\x07mobends\x040.24\x12mob_grinding_utils\x060.3.13\tmobstages\x062.0.13\x10modularmachinery\x051.9.5\x08morpheus\x0e1.12.2-3.5.106\x07mputils\x051.5.6\x07mpbasic\x051.4.7\x10multiblockstages\x051.1.1\x0fmundaneredstone\x051.1.4\x13mysticalagriculture\x051.7.3\x14mysticalagradditions\x051.3.1\x0fmystagradcompat\x031.2\x0enaturescompass\x051.5.1\x03nex\x052.0.8\x0enoworldgen5you\x051.0.6\tnutrition\x053.5.0\tsamsocean\x051.0.1\roreexcavation\x071.4.137\roeintegration\x052.3.4\torestages\x062.0.37\noverloaded\x060.0.53\x0cpickletweaks\x052.1.1\x07placebo\x051.5.1\x0cplayerbosses\x031.0\x0bplayerskins\x051.0.4\x0epneumaticcraft\x101.12.2-0.8.4-303\rpoweradapters\x051.0.9\x08prestige\x061.1.30\x0cprimalchests\x051.0.3\x06rustic\x051.1.0\x0fthebetweenlands\x053.4.6\x06primal\x070.6.104\x11progressiontweaks\r1.12.2-0.3.40\x0bprospectors\x051.0.2\nreborncore\n3.13.5.421\x0equantumstorage\x054.6.8\x0equickleafdecay\x051.2.4\x0brangedpumps\x030.5\trealdrops\x061.2.14\rrebornstorage\x051.0.0\x0crecipestages\x051.1.1\nreccomplex\x051.4.7\x14refinedstorageaddons\x050.4.3\x07rftools\x047.61\x0erftoolscontrol\x051.9.3\nroadrunner\x051.0.1\x07sampler\x041.80\tscannable\x081.6.3.24\tsevtweaks\x080.2.3-20\x0esev_tweaks_npc\x050.0.4\rsimpleautorun\n1.12.1-1.2\x10simplegenerators\x0f1.12.2-2.0.19.1\x0estoragenetwork\x051.5.7\x0csimplyarrows\x051.0.4\x0espartanshields\x051.4.1\x10spatialservermod\x031.3\x0bstevescarts\n2.4.29.132\x03stg\x0c1.12.2-1.2.3\x0estoragedrawers\n1.12-5.3.5\x07streams\x050.4.4\x05sasit\x061.1.14\x11supersoundmuffler\x081.0.2.10\ttallgates\x051.0.0\x07beneath\x051.5.0\x0ethirstybottles\x051.1.4\x0btcomplement\n${version}\x0ctinkerstages\x062.0.17\x12tinkertoolleveling\x181.12.2-1.1.0.DEV.b23e769\x0ftogetherforever\x051.0.2\ttombstone\x053.3.2\x07totemic\r1.12.2-0.11.5\rtothebatpoles\x0e1.12.2-1.1.0.1\rtranslocators\x082.5.1.77\x07triumph\x063.13.0\x0ftrumpetskeleton\x0c1.12-1.0.2.1\ntumbleweed\n1.11-0.4.5\x06uppers\x050.0.6\x12universalmodifiers\x0f1.12.2-1.0.16.1\x02vc\x065.9.13\x03vtt\x050.6.4\x07waddles\x050.6.0\x0bwailastages\x062.0.23\twanionlib\n1.12.2-2.2\x13watercontrolextreme\x051.0.0\rwaterstrainer\x053.2.0\x05wawla\x072.5.270\x0eweirdinggadget\x031.0\twildcrops\x051.0.1\x0ewitherskelefix\x052.6.3\x06wopper\x071.12-r5\rworldstripper\x0c1.6.0-1.12.2\x04xnet\x051.7.6\x04ynot\x050.2.3\x05yoyos\x0f1.12.2-1.2.2.20\tzenstages\x080.4.1-19\x0bprimal_tech\x050.3.3\x17teslacorelib_registries\x061.0.15'
</details>

The server also logs an error during decompression:

[07:25:54] [Netty Server IO #8/INFO] [FML]: Client protocol version 2
[07:25:54] [Netty Server IO #8/ERROR] [FML]: NetworkDispatcher exception
io.netty.handler.codec.DecoderException: java.util.zip.DataFormatException: incorrect header check
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:442) ~[minecraft_server.1.12.2.jar:?]
...
Caused by: java.util.zip.DataFormatException: incorrect header check
	at java.util.zip.Inflater.inflateBytes(Native Method) ~[?:1.8.0_131]

--- a/src/protocol/mod.rs
+++ b/src/protocol/mod.rs
@@ -1020,6 +1020,12 @@ impl Conn {
             VarInt(uncompressed_size as i32).write_to(&mut new)?;
             let mut write = ZlibEncoder::new(io::Cursor::new(buf), Compression::default());
             write.read_to_end(&mut new)?;
+            let network_debug = unsafe { NETWORK_DEBUG };
+            if network_debug {
+                println!("Compressed {} bytes to {} since > threshold {}, new={:?}",
+                         uncompressed_size, new.len(), self.compression_threshold,
+                         new);
+            }
             buf = new;
         }
 

--- a/src/protocol/mod.rs
+++ b/src/protocol/mod.rs
@@ -972,7 +972,6 @@ pub struct Conn {
     cipher: Option<Aes128Cfb>,
 
     compression_threshold: i32,
-    compression_read: Option<ZlibDecoder<io::Cursor<Vec<u8>>>>,
 }
 
 impl Conn {
@@ -999,7 +998,6 @@ impl Conn {
             protocol_version,
             cipher: Option::None,
             compression_threshold: -1,
-            compression_read: Option::None,
         })
     }
 
@@ -1022,7 +1020,7 @@ impl Conn {
             write.read_to_end(&mut new)?;
             let network_debug = unsafe { NETWORK_DEBUG };
             if network_debug {
-                println!("Compressed {} bytes to {} since > threshold {}, new={:?}",
+                println!("Compressed for sending {} bytes to {} since > threshold {}, new={:?}",
                          uncompressed_size, new.len(), self.compression_threshold,
                          new);
             }
@@ -1039,6 +1037,7 @@ impl Conn {
     }
 
     pub fn read_packet(&mut self) -> Result<packet::Packet, Error> {
+        let network_debug = unsafe { NETWORK_DEBUG };
         let len = VarInt::read_from(self)?.0 as usize;
         let mut ibuf = vec![0; len];
         self.read_exact(&mut ibuf)?;
@@ -1046,16 +1045,16 @@ impl Conn {
         let mut buf = io::Cursor::new(ibuf);
 
         if self.compression_threshold >= 0 {
-            if self.compression_read.is_none() {
-                self.compression_read = Some(ZlibDecoder::new(io::Cursor::new(Vec::new())));
-            }
             let uncompressed_size = VarInt::read_from(&mut buf)?.0;
             if uncompressed_size != 0 {
                 let mut new = Vec::with_capacity(uncompressed_size as usize);
                 {
-                    let reader = self.compression_read.as_mut().unwrap();
-                    reader.reset(buf);
+                    let mut reader = ZlibDecoder::new(buf);
                     reader.read_to_end(&mut new)?;
+                }
+                if network_debug {
+                    println!("Decompressed threshold={} len={} uncompressed_size={} to {} bytes",
+                        self.compression_threshold, len, uncompressed_size, new.len());
                 }
                 buf = io::Cursor::new(new);
             }
@@ -1066,8 +1065,6 @@ impl Conn {
             Direction::Clientbound => Direction::Serverbound,
             Direction::Serverbound => Direction::Clientbound,
         };
-
-        let network_debug = unsafe { NETWORK_DEBUG };
 
         if network_debug {
             println!("about to parse id={:x}, dir={:?} state={:?}", id, dir, self.state);
@@ -1274,7 +1271,6 @@ impl Clone for Conn {
             protocol_version: self.protocol_version,
             cipher: Option::None,
             compression_threshold: self.compression_threshold,
-            compression_read: Option::None,
         }
     }
 }

--- a/src/protocol/mod.rs
+++ b/src/protocol/mod.rs
@@ -1016,31 +1016,16 @@ impl Conn {
             0
         };
         if self.compression_threshold >= 0 && buf.len() as i32 > self.compression_threshold {
-            println!("Compressing since threshold={} > len={}", self.compression_threshold, buf.len());
             if self.compression_write.is_none() {
                 self.compression_write = Some(ZlibEncoder::new(io::Cursor::new(Vec::new()), Compression::default()));
-                println!("Initialized self.compression_write");
-            } else {
-                println!("Not initializing self.compression_write, was already");
             }
-
             extra = 0;
             let uncompressed_size = buf.len();
             let mut new = Vec::new();
-            //VarInt(uncompressed_size as i32).write_to(&mut new)?;
+            VarInt(uncompressed_size as i32).write_to(&mut new)?;
             let write = self.compression_write.as_mut().unwrap();
-            println!("Compressing buf={:?}", buf);
             write.reset(io::Cursor::new(buf));
             write.read_to_end(&mut new)?;
-            println!("Compressed bytes={:?}", new);
-            println!("About to decompress");
-            {
-                let mut reader = ZlibDecoder::new(&new[..]);
-                let mut check = Vec::new();
-                reader.read_to_end(&mut check).unwrap();
-                println!("Decompressed bytes={:?}", check);
-            }
-
             buf = new;
         }
 

--- a/src/protocol/mod.rs
+++ b/src/protocol/mod.rs
@@ -1016,15 +1016,18 @@ impl Conn {
             0
         };
         if self.compression_threshold >= 0 && buf.len() as i32 > self.compression_threshold {
+            /*
             if self.compression_write.is_none() {
                 self.compression_write = Some(ZlibEncoder::new(io::Cursor::new(Vec::new()), Compression::default()));
             }
+            */
             extra = 0;
             let uncompressed_size = buf.len();
             let mut new = Vec::new();
             VarInt(uncompressed_size as i32).write_to(&mut new)?;
-            let write = self.compression_write.as_mut().unwrap();
-            write.reset(io::Cursor::new(buf));
+            //let write = self.compression_write.as_mut().unwrap();
+            //write.reset(io::Cursor::new(buf));
+            let mut write = ZlibEncoder::new(buf.as_slice(), Compression::default());
             write.read_to_end(&mut new)?;
             buf = new;
         }

--- a/src/protocol/mod.rs
+++ b/src/protocol/mod.rs
@@ -1016,16 +1016,31 @@ impl Conn {
             0
         };
         if self.compression_threshold >= 0 && buf.len() as i32 > self.compression_threshold {
+            println!("Compressing since threshold={} > len={}", self.compression_threshold, buf.len());
             if self.compression_write.is_none() {
                 self.compression_write = Some(ZlibEncoder::new(io::Cursor::new(Vec::new()), Compression::default()));
+                println!("Initialized self.compression_write");
+            } else {
+                println!("Not initializing self.compression_write, was already");
             }
+
             extra = 0;
             let uncompressed_size = buf.len();
             let mut new = Vec::new();
-            VarInt(uncompressed_size as i32).write_to(&mut new)?;
+            //VarInt(uncompressed_size as i32).write_to(&mut new)?;
             let write = self.compression_write.as_mut().unwrap();
+            println!("Compressing buf={:?}", buf);
             write.reset(io::Cursor::new(buf));
             write.read_to_end(&mut new)?;
+            println!("Compressed bytes={:?}", new);
+            println!("About to decompress");
+            {
+                let mut reader = ZlibDecoder::new(&new[..]);
+                let mut check = Vec::new();
+                reader.read_to_end(&mut check).unwrap();
+                println!("Decompressed bytes={:?}", check);
+            }
+
             buf = new;
         }
 

--- a/src/protocol/mod.rs
+++ b/src/protocol/mod.rs
@@ -973,7 +973,6 @@ pub struct Conn {
 
     compression_threshold: i32,
     compression_read: Option<ZlibDecoder<io::Cursor<Vec<u8>>>>,
-    compression_write: Option<ZlibEncoder<io::Cursor<Vec<u8>>>>,
 }
 
 impl Conn {
@@ -1001,7 +1000,6 @@ impl Conn {
             cipher: Option::None,
             compression_threshold: -1,
             compression_read: Option::None,
-            compression_write: Option::None,
         })
     }
 
@@ -1016,18 +1014,11 @@ impl Conn {
             0
         };
         if self.compression_threshold >= 0 && buf.len() as i32 > self.compression_threshold {
-            /*
-            if self.compression_write.is_none() {
-                self.compression_write = Some(ZlibEncoder::new(io::Cursor::new(Vec::new()), Compression::default()));
-            }
-            */
             extra = 0;
             let uncompressed_size = buf.len();
             let mut new = Vec::new();
             VarInt(uncompressed_size as i32).write_to(&mut new)?;
-            //let write = self.compression_write.as_mut().unwrap();
-            //write.reset(io::Cursor::new(buf));
-            let mut write = ZlibEncoder::new(buf.as_slice(), Compression::default());
+            let mut write = ZlibEncoder::new(io::Cursor::new(buf), Compression::default());
             write.read_to_end(&mut new)?;
             buf = new;
         }
@@ -1278,7 +1269,6 @@ impl Clone for Conn {
             cipher: Option::None,
             compression_threshold: self.compression_threshold,
             compression_read: Option::None,
-            compression_write: Option::None,
         }
     }
 }

